File: Sources/OpenZCineCore/ReconnectBackoff.swift
import Foundation

/// Jittered exponential backoff schedule for reconnect / live-view-restart retries.
///
/// Plain doubling makes every client (and every retry within one client) wake at the same instants,
/// so a flaky access point gets hit by synchronized reconnect bursts. Spreading each delay by a
/// random fraction de-correlates the attempts — the standard mitigation for reconnect storms.
///
/// The jitter sample is injected (`0...1`) rather than drawn internally so the schedule is pure and
/// unit-testable: pass `Double.random(in: 0...1)` at the call site, or `0.5` for the midpoint.
public struct ReconnectBackoff: Sendable, Equatable {
    /// Delay used for the very first retry (attempt 0), before any exponential growth.
    public var baseSeconds: Double
    /// Hard ceiling on the returned delay, applied before and after jitter.
    public var maxSeconds: Double
    /// Per-attempt growth factor (2 = double each attempt).
    public var multiplier: Double
    /// Proportional spread applied to each delay, in `0...1` (0.3 = ±30%).
    public var jitterFraction: Double

    public init(
        baseSeconds: Double = 0.5,
        maxSeconds: Double = 30,
        multiplier: Double = 2,
        jitterFraction: Double = 0.3
    ) {
        self.baseSeconds = baseSeconds
        self.maxSeconds = maxSeconds
        self.multiplier = multiplier
        self.jitterFraction = jitterFraction
    }

    /// The delay (seconds) to wait before retry `attempt` (0 = first retry, growing from there).
    ///
    /// - Parameters:
    ///   - attempt: zero-based retry index; negatives are treated as 0.
    ///   - jitter: a sample in `0...1` (clamped); 0 = lower bound, 0.5 = un-jittered, 1 = upper bound.
    public func delaySeconds(forAttempt attempt: Int, jitter: Double) -> Double {
        let exponent = Double(max(0, attempt))
        let capped = min(maxSeconds, baseSeconds * pow(multiplier, exponent))
        let clampedJitter = min(1, max(0, jitter))
        // Map the 0...1 sample onto ±jitterFraction around the un-jittered delay.
        let signedSpread = (clampedJitter * 2 - 1) * jitterFraction
        let jittered = capped * (1 + signedSpread)
        return min(maxSeconds, max(0, jittered))
    }
}

File: Apps/Android/app/src/main/kotlin/com/opencapture/openzcine/transport/ReconnectBackoff.kt
package com.opencapture.openzcine.transport

import kotlin.math.max
import kotlin.math.pow
import kotlin.math.roundToLong

/**
 * Jittered exponential backoff for reconnect retries — the Kotlin twin of the
 * shared core's `ReconnectBackoff` (`Sources/OpenZCineCore/ReconnectBackoff.swift`).
 *
 * Plain doubling wakes every retry at the same instants and hammers a flaky
 * access point with synchronized bursts; spreading each delay by a random
 * fraction de-correlates them. The jitter sample is injected (`0.0..1.0`)
 * rather than drawn internally so the schedule stays pure and unit-testable.
 *
 * @property baseMillis Delay for the very first retry (attempt 0).
 * @property maxMillis Hard ceiling, applied before and after jitter.
 * @property multiplier Per-attempt growth factor (2.0 = double each attempt).
 * @property jitterFraction Proportional spread in `0.0..1.0` (0.3 = ±30%).
 */
data class ReconnectBackoff(
    val baseMillis: Long = 500,
    val maxMillis: Long = 30_000,
    val multiplier: Double = 2.0,
    val jitterFraction: Double = 0.3,
) {
    /**
     * The delay in milliseconds before retry [attempt] (0 = first retry;
     * negatives are treated as 0). [jitter] is a sample in `0.0..1.0`
     * (clamped): 0 = lower bound, 0.5 = un-jittered, 1 = upper bound.
     */
    fun delayMillis(attempt: Int, jitter: Double): Long {
        val exponent = max(0, attempt).toDouble()
        val capped = minOf(maxMillis.toDouble(), baseMillis * multiplier.pow(exponent))
        val clampedJitter = jitter.coerceIn(0.0, 1.0)
        val signedSpread = (clampedJitter * 2 - 1) * jitterFraction
        val jittered = capped * (1 + signedSpread)
        return jittered.coerceIn(0.0, maxMillis.toDouble()).roundToLong()
    }
}
